### PR TITLE
Barracuda integration into ML-Agents

### PR DESCRIPTION
**Pull request by mantasp**
This is preview stage of the Barracuda integration into ML-Agents. 
It should work right out of the box. Just open test scene and hit play.
There will be code cleanup round on top of current work, but it shouldn't bring any dramatic changes.

### CODE REVIEW

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Plugins/iOS.meta [hunk 1]
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Plugins/iOS.meta]: There does not seems to be an asset associated with this meta file

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta [hunk 2]
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta]: I think we should remove the Tools folder now that we included the python converter into the python side of mlagents

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+                    "Action Size of the model does not match. " +
+                    "The BrainParameters expect {0} but the model contains {1}.",
+                    bpActionSize, modelActionSize);
+            }
+            return null;
+        }
+    }
+}
+
+public class BarracudaUtils
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs]: This should be in its own script

File: UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs [hunk 4]
```diff
@@ -1,9 +1,13 @@
-﻿using System;
+﻿#define ENABLE_BARRACUDA
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs]: We should remove TF# and these define symbols before merging

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Resources/BarracudaReferenceImpl.compute [hunk 5]
```diff
@@ -0,0 +1,863 @@
+﻿#pragma kernel Dense
```
Comment (vladimiroster) [inline UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Resources/BarracudaReferenceImpl.compute]: Are you shipping with both the DLL and the source?

File: UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs (hunk 4 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs]: We will leave the TF# code in place for the next release in order to transition.

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Resources/BarracudaReferenceImpl.compute (hunk 5 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Resources/BarracudaReferenceImpl.compute]: The compute files cannot be put in a DLL.

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta (hunk 2 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta]: The Tools folder will be left here for iteration and will be removed before the release. Leaving this comment unresolved so we do not forget

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Plugins/iOS.meta (hunk 1 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Barracuda/Plugins/iOS.meta]: fixed, I committed missing iOS plugin

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta (hunk 2 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta]: Waiting for Rej to integrate your changes into Barracuda distribution then will push new version of Barracuda and remove this folder

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs (hunk 3 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs]: I expect these to live for one release only, after we move to Barracuda Tensor across the board most of this code will disappear..

**Comment by mantasp** [comment]
@awjuliani thanks for trying it! The main purpose was just give you heads up and get a feedback if it works for you.

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs [hunk 6]
```diff
@@ -125,14 +168,8 @@ public class PreviousActionInputGenerator : TensorGenerator.Generator
     {
         public void Generate(Tensor tensor, int batchSize, Dictionary<Agent, AgentInfo> agentInfo)
         {
-            if (tensor.ValueType != Tensor.TensorType.Integer)
-            {
-                throw new NotImplementedException(
-                    "Previous Action Inputs are only valid for discrete control");
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs]: Why was this removed ? It breaks the Unit Tests

File: ml-agents/tests/trainers/test_barracuda_converter.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+import mlagents.trainers.tensorflow_to_barracuda as tf2bc
+
+
+def test_barracuda_converter():
+    path_prefix = os.path.dirname(os.path.abspath(__file__))
+    tmpfile = os.path.join(tempfile._get_default_tempdir(), next(tempfile._get_candidate_names()) + '.nn')
+
+    # make sure file doesn't exist before conversion
+    assert (not os.path.isfile(tmpfile))
```
Comment (vincentpierre) [inline ml-agents/tests/trainers/test_barracuda_converter.py]: This could break if the line 23 is not called. (Because of errors in the asset calls). This test would fail thereafter. I would remove the file if exists at the beginning of the test. Also, Some comments on this method.

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs [hunk 8]
```diff
@@ -0,0 +1,615 @@
...
+        /// <param name="visObsIndex"> The index of the visual observation.</param>
+        /// <returns>If the Check failed, returns a string containing information about why the
+        /// check failed. If the check passed, returns null.</returns>
+        private string CheckVisualObsShape(Tensor tensor, int visObsIndex)
+        {
+            var resolutionBp = _brainParameters.cameraResolutions[visObsIndex];
+            var widthBp = resolutionBp.width;
+            var heightBp = resolutionBp.height;
+            var pixelBp = resolutionBp.blackAndWhite ? 1 : 3;
+            var widthT = tensor.Shape[1];
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs]: widthT and heightT must be switched. It was a bug that was fixed in v0.6.0a

File: UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta (hunk 2 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Plugins/Barracuda.Core/Tools.meta]: Tools are gone by now

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs (hunk 6 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs]: Barracuda doesn't support Integer tensors, all tensors currently are float32

File: ml-agents/tests/trainers/test_barracuda_converter.py (hunk 7 above)
Comment (mantasp) [reply ml-agents/tests/trainers/test_barracuda_converter.py]: will fix it

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs (hunk 8 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs]: will fix it

File: ml-agents/tests/trainers/test_barracuda_converter.py (hunk 7 above)
Comment (mantasp) [reply ml-agents/tests/trainers/test_barracuda_converter.py]: fixed

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs (hunk 8 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs]: fixed

File: UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs (hunk 6 above)
Comment (mantasp) [reply UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/GeneratorImpl.cs]: I changed Edit mode GeneratePreviousActionInput test to match new logic
